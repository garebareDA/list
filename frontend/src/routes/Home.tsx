import React, { useState } from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';

import Chat from '../components/Chat';

function Home() {
    const [room, setRoom] = useState("");
    const history = useHistory();

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setRoom(event.target.value);
    }

    const onClickRoom = () => {
        fetch('/api/group', {
            method: 'POST',
            mode: 'cors',
            cache: 'no-cache',
            headers: {
                'Content-Type': 'application/json'
            },
            redirect: 'follow', 
            referrerPolicy: 'no-referrer',
            body: JSON.stringify({
                name:room,
            })
        }).then((res) => {
            res.json().then((json) => {
                history.push("/group/" + json.id);
            });
        })
    }

    return (
        <div>
            <AutoCenter>
                <TitleText> ListChat </TitleText>
                <TextArea maxLength={32} placeholder="ルーム名" onChange={handleChange} ></TextArea>
                <RoomCreateButton onClick={onClickRoom} disabled={room == ""}>ルームを作成</RoomCreateButton>
            </AutoCenter>
            <PreviewChat>
                <Chat me={{ img: "https://pbs.twimg.com/profile_images/1437747193234944002/a9f_91G8_400x400.jpg", name: "ガレバレ", message: "たぶん、リアルタイムなチャットです" }} sendTo={null} />
                <Chat me={{ img: "https://pbs.twimg.com/profile_images/1437747193234944002/a9f_91G8_400x400.jpg", name: "ガレバレ", message: "たぶん、一覧で表示されます" }} sendTo={null} />
                <Chat me={{ img: "https://pbs.twimg.com/profile_images/1437747193234944002/a9f_91G8_400x400.jpg", name: "ガレバレ", message: "たぶん、みんなで使えます" }} sendTo={null} />
            </PreviewChat>
        </div>
    );
}

export default Home;

const TitleText = styled.div`
    font-size: 32px;
    color: white;
    padding:32px;
`

const TextArea = styled.input`
    box-shadow: 0 1px 25px 0 rgba(0, 0, 0, .5);
    border-radius: 1em;
    padding: 1em;
    background-color: snow;
    width: 50%;
    max-width: 512px;
    min-width:128px;
    box-sizing: border-box;
    max-height: 32px;
    font-size: 100%;
    line-height: 1.2;
    outline: none;
    resize:none;
    border: none;
`

const RoomCreateButton = styled.button`
    box-shadow: 0 1px 25px 0 rgba(0, 0, 0, .5);
    outline: none;
    border: none;
    border-radius: 1em;
    width:128px;
    height:32px;
    color:white;
    background-color: #005eff;
    margin: 8px;

    :hover {
        background-color: #054ecb;
    };

    :active {
        background-color: #005eff;
    };

    :disabled {
        color: gray;
        background-color: #bababa;
    }
`

const AutoCenter = styled.div`
    text-align:center;
    padding-top:128px;
    padding-bottom: 80px;
`

const PreviewChat = styled.div`
    width:auto;
    height:auto;
    margin: 0 auto;
    display:flex;
    flex-wrap:wrap;
    justify-content: center;
    align-items: center;
`
