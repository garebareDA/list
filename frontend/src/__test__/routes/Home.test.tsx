import { render, fireEvent } from '@testing-library/react'
import fetchMock, { MockResponseInit } from 'jest-fetch-mock';
import Home from '../../routes/Home'

jest.mock('socket.io-client')
jest.mock('react-router', () => ({
    ...jest.requireActual('react-router'),
    useHistory: () => ({ push: jest.fn() })
}));

test('HomeRender', () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot()
});

test('onClickRoom', () => {
    const value = "room";
    fetchMock.enableMocks();
    fetchMock.mockIf('/api/group', (request: Request): Promise<MockResponseInit> => {
        request.json().then((json) => { expect(json.name).toBe(value) });

        const res = JSON.stringify(
            {
                name: "room",
                id: "123",
            }
        );

        return new Promise((resolve, reject) => {
            resolve({
                body: res
            })
        })
    });

    const { getByText, getByPlaceholderText, debug } = render(<Home />);
    const roomCreateButton = getByText("ルームを作成");
    const input = getByPlaceholderText("ルーム名");

    fireEvent.change(input, { target: { value: value } });
    expect((input as HTMLInputElement).value).toBe(value);

    fireEvent.click(roomCreateButton);
});


