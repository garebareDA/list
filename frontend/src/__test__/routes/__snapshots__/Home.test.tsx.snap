// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeRender 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-kEjbQP cvYYPK"
    >
      <div
        class="sc-fubCzh dFCHyE"
      >
         ListChat 
      </div>
      <input
        class="sc-pGacB khYXTb"
        maxlength="32"
        placeholder="ルーム名"
      />
      <button
        class="sc-jrAFXE bEXVrr"
        disabled=""
      >
        ルームを作成
      </button>
    </div>
    <div
      class="sc-iqHYmW gpXrTA"
    >
      <div
        class="sc-bdfBQB gcvDuM"
      >
        <div
          class="sc-gsTEea fIwExb"
        >
          <img
            alt="画像"
            class="sc-dlfnuX ctYUkh"
            src=""
          />
          <div
            class="sc-hKgJUU hJkrTx"
          >
            ガレバレ
          </div>
        </div>
        <div
          class="sc-eCstlR gAGJiT"
        >
          たぶん、リアルタイムなチャットです
        </div>
        <div />
      </div>
      <div
        class="sc-bdfBQB gcvDuM"
      >
        <div
          class="sc-gsTEea fIwExb"
        >
          <img
            alt="画像"
            class="sc-dlfnuX ctYUkh"
            src=""
          />
          <div
            class="sc-hKgJUU hJkrTx"
          >
            ガレバレ
          </div>
        </div>
        <div
          class="sc-eCstlR gAGJiT"
        >
          たぶん、一覧で表示されます
        </div>
        <div />
      </div>
      <div
        class="sc-bdfBQB gcvDuM"
      >
        <div
          class="sc-gsTEea fIwExb"
        >
          <img
            alt="画像"
            class="sc-dlfnuX ctYUkh"
            src=""
          />
          <div
            class="sc-hKgJUU hJkrTx"
          >
            ガレバレ
          </div>
        </div>
        <div
          class="sc-eCstlR gAGJiT"
        >
          たぶん、みんなで使えます
        </div>
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
